File: src/main/java/net/softbell/bsh/controller/view/ActionView.java
package net.softbell.bsh.controller.view;

import java.util.List;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;

import lombok.AllArgsConstructor;
import net.softbell.bsh.domain.entity.NodeAction;
import net.softbell.bsh.domain.entity.NodeActionItem;
import net.softbell.bsh.domain.entity.NodeItem;
import net.softbell.bsh.dto.request.IotActionDto;
import net.softbell.bsh.iot.service.v1.IotActionServiceV1;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 모니터 뷰 컨트롤러
 */
@AllArgsConstructor
@Controller
@RequestMapping("/action")
public class ActionView
{
	// Global Field
	private final String G_BASE_PATH = "services/general";
	private final IotActionServiceV1 iotActionService;
	
	@GetMapping()
    public String dispIndex(Model model, Authentication auth)
	{
		// Field
		List<NodeAction> listNodeAction;
		
		// Init
		listNodeAction = iotActionService.getAllNodeActions(auth);
		
		// Process
		model.addAttribute("listNodeAction", listNodeAction);
		
		// Return
        return G_BASE_PATH + "/Action";
    }
	
	@GetMapping("/{id}")
    public String dispAction(Model model, Authentication auth, @PathVariable("id") long actionId)
	{
		// Field
		NodeAction nodeAction;
		List<NodeItem> listNodeItem;
		
		// Init
		nodeAction = iotActionService.getNodeAction(auth, actionId);
		listNodeItem = iotActionService.getAvailableNodeItem(auth);
		
		for (NodeActionItem actionItem : nodeAction.getNodeActionItems())
			listNodeItem.remove(actionItem.getNodeItem());
		
		// Process
		model.addAttribute("actionInfo", nodeAction);
		model.addAttribute("listNodeItemAct", nodeAction.getNodeActionItems());
		model.addAttribute("listNodeItem", listNodeItem);
		
		// Return
        return G_BASE_PATH + "/ActionModify";
    }
	
	@GetMapping("/create")
    public String dispCreate(Model model, Authentication auth)
	{
		// Field
		List<NodeItem> listNodeItem;
		
		// Init
		listNodeItem = iotActionService.getAvailableNodeItem(auth);
		
		// Process
		model.addAttribute("listNodeItem", listNodeItem);
		
		// Return
        return G_BASE_PATH + "/ActionCreate";
    }
	
	@PostMapping("/create")
    public String procCreate(Model model, Authentication auth,
    						IotActionDto iotActionDto)
	{
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotActionService.createAction(auth, iotActionDto);
		
		// Return
		if (isSuccess)
			return "redirect:/action";
		else
			return "redirect:/action?error";
    }
	
	@PostMapping("/modify/{id}")
    public String procModify(Model model, Authentication auth,
    						@PathVariable("id") long actionId,
    						IotActionDto iotActionDto)
	{
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotActionService.modifyAction(auth, actionId, iotActionDto);
		
		// Return
		if (isSuccess)
			return "redirect:/action/";
		else
			return "redirect:/action/modify/" + actionId + "?error";
    }
	
	@PostMapping("/delete/{id}")
    public String procDelete(Model model, Authentication auth, @PathVariable("id") long actionId)
	{
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotActionService.deleteAction(auth, actionId);
		
		// Return
		if (isSuccess)
			return "redirect:/action";
		else
			return "redirect:/action/modify/" + actionId + "?error";
    }
}

File: src/main/kotlin/net/softbell/bsh/controller/view/general/ActionView.kt
package net.softbell.bsh.controller.view.general

import net.softbell.bsh.domain.entity.Member
import net.softbell.bsh.domain.entity.NodeItem
import net.softbell.bsh.dto.request.IotActionDto
import net.softbell.bsh.dto.view.general.ActionInfoCardDto
import net.softbell.bsh.iot.service.v1.IotActionServiceV1
import net.softbell.bsh.service.CenterService
import net.softbell.bsh.service.InterlockService
import net.softbell.bsh.service.ViewDtoConverterService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import javax.servlet.http.HttpServletRequest

/**
 * @author : Bell(bell@softbell.net)
 * @description : 모니터 뷰 컨트롤러
 */
@Controller
@RequestMapping("/action")
class ActionView {
    // Global Field
    private val G_INDEX_REDIRECT_URL: String = "redirect:/"

    @Autowired private lateinit var viewDtoConverterService: ViewDtoConverterService
    @Autowired private lateinit var iotActionService: IotActionServiceV1
    @Autowired private lateinit var centerService: CenterService
    @Autowired private lateinit var interlockService: InterlockService

    @GetMapping
    fun dispIndex(model: Model, @AuthenticationPrincipal member: Member): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val listNodeAction = iotActionService.getPrivilegesNodeActions(member)

        // Process
        model.addAttribute("listCardActions", viewDtoConverterService.convActionSummaryCards(listNodeAction))

        // Return
        return "services/general/Action"
    }

    @GetMapping("/{id}")
    fun dispAction(model: Model, @AuthenticationPrincipal member: Member, request: HttpServletRequest,
                   @PathVariable("id") actionId: Long): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val nodeAction = iotActionService.getPrivilegesNodeAction(member, actionId)
        val listNodeItem: MutableList<NodeItem> = iotActionService.getAvailableNodeItem(member) as MutableList<NodeItem>
        val listMemberInterlockToken = interlockService.getAllTokens(member)
        var baseUrl = request.requestURL.toString()

        baseUrl = baseUrl.substring(0, baseUrl.indexOf("/", 8))
        if (nodeAction != null)
            for (actionItem in nodeAction.nodeActionItems)
                listNodeItem.remove(actionItem.nodeItem)

        // Process
        model.addAttribute("cardActionInfo", nodeAction?.let { ActionInfoCardDto(it) })
        model.addAttribute("listCardInterlocks", viewDtoConverterService.convActionInterlockTokenCards(listMemberInterlockToken))
        if (nodeAction != null)
            model.addAttribute("listCardItems", viewDtoConverterService.convActionItemCards(nodeAction.nodeActionItems))
        model.addAttribute("baseURL", baseUrl)

        // Return
        return "services/general/ActionInfo"
    }

    @GetMapping("/modify/{id}")
    fun dispActionModify(model: Model, @AuthenticationPrincipal member: Member,
                         @PathVariable("id") actionId: Long): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val nodeAction = iotActionService.getPrivilegesNodeAction(member, actionId)
        val listNodeItem = iotActionService.getAvailableNodeItem(member) as MutableList<NodeItem>

        if (nodeAction != null)
            for (actionItem in nodeAction.nodeActionItems)
                listNodeItem.remove(actionItem.nodeItem)

        // Process
        model.addAttribute("cardActionInfo", nodeAction?.let { ActionInfoCardDto(it) })
        if (nodeAction != null)
            model.addAttribute("listCardItemActives", viewDtoConverterService.convActionItemCards(nodeAction.nodeActionItems))
        model.addAttribute("listCardItems", viewDtoConverterService.convActionItemCards(listNodeItem))

        // Return
        return "services/general/ActionModify"
    }

    @GetMapping("/create")
    fun dispCreate(model: Model, @AuthenticationPrincipal member: Member): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val listNodeItem = iotActionService.getAvailableNodeItem(member)

        // Process
        model.addAttribute("listCardItems", viewDtoConverterService.convActionItemCards(listNodeItem))

        // Return
        return "services/general/ActionCreate"
    }

    @PostMapping("/create")
    fun procCreate(model: Model, @AuthenticationPrincipal member: Member,
                   iotActionDto: IotActionDto): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotActionService.createAction(member, iotActionDto)

        // Return
        return if (isSuccess)
            "redirect:/action"
        else
            "redirect:/action?error"
    }

    @PostMapping("/modify/{id}")
    fun procModify(model: Model, @AuthenticationPrincipal member: Member,
                   @PathVariable("id") actionId: Long,
                   iotActionDto: IotActionDto): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotActionService.modifyPrivilegesAction(member, actionId, iotActionDto)

        // Return
        return if (isSuccess)
            "redirect:/action"
        else
            "redirect:/action/modify/$actionId?error"
    }

    @PostMapping("/delete/{id}")
    fun procDelete(model: Model, @AuthenticationPrincipal member: Member,
                   @PathVariable("id") actionId: Long): String {
        // Exception
        if (centerService.setting.iotAction != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotActionService.deletePrivilegesAction(member, actionId)

        // Return
        return if (isSuccess)
            "redirect:/action"
        else
            "redirect:/action/modify/$actionId?error"
    }
}